### ISSUE TITLE
Torch Multi Head Attention

### ISSUE THREAD

**Issue by grazder** [NVIDIA/NeMo#6257]
**Is your feature request related to a problem? Please describe.**

I've seen that you are using custom implementation of MHA block - https://github.com/NVIDIA/NeMo/blob/da6d825b8c2fec0a3b8e0d2c33a7600f242db4c1/nemo/collections/asr/parts/submodules/multi_head_attention.py#L108

I think it would be better to use `torch.nn.MultiheadAttention` which is meant to improve the performance of this block (there is FlashAttention now) and torch continues to maintain and update this block while you're doing basic implementation of it.

**Describe the solution you'd like**

I'd like to see an update of models with torch implementation of MHA

https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html

What do you think about it?

**Comment by titu1994** [comment]
The pytorch MHA implementation is specific to absolutele pos embedding MHA used commonly in NLP. In ASR, we use relative positional embedding for transformer blocks - https://github.com/NVIDIA/NeMo/blob/da6d825b8c2fec0a3b8e0d2c33a7600f242db4c1/nemo/collections/asr/parts/submodules/multi_head_attention.py#L211

If you look at pytorch code, the mask shape is not directly same, but can be made similar (we use BxT1xT2, they use BxHxL1xL2) 

@SeanNaren can you try it to see if it works ?

**Comment by titu1994** [comment]
That's pytorch 2.0, which we don't currently test against but will soon. Even then, it will not be the minimum version for some time so we cannot switch to it right now

**Comment by desh2608** [comment]
Another constraint at the moment is that the implementation using flash attention does not support arbitrary attention masks (https://github.com/pytorch/pytorch/pull/81434#issuecomment-1478759658).

**Comment by vadimkantorov** [comment]
I guess we should also raise this issue in xformers as it might be faster to evolve

**Comment by SeanNaren** [comment]
@vadimkantorov we had an issue in xFormer: https://github.com/facebookresearch/xformers/issues/558

It was closed after support for an additive bias was added directly into xFormer. However we still need to create a model that uses an additive bias as apposed to relative bias/shift to be able to use it unfortunately...

**Comment by raman-r4334** [comment]
Hi @SeanNaren , 

Please let me know if you make `additive bias` work for conformer based models

**Comment by vadimkantorov** [comment]
These days PyTorch core SDPA should better support attn_masks / relpos, I wonder if this should be reconsidered or submit a feature request to PyTorch if some specific relpos fusion need to be supported

**Comment by vadimkantorov** [comment]
@SeanNaren btw has nvidia tried using RoPE instead of relpos for conformer? in theory it should enable use of SDPA: https://pytorch.org/blog/pytorch-compile-to-speed-up-inference/ (and maybe torch.compile even during training)

**Comment by vadimkantorov** [comment]
@SeanNaren couldn't we replace

```python
            q, k, v = self.forward_qkv(query, key, value)
            q = q.transpose(1, 2)  # (batch, time1, head, d_k)

            n_batch_pos = pos_emb.size(0)
            p = self.linear_pos(pos_emb).view(n_batch_pos, -1, self.h, self.d_k)
            p = p.transpose(1, 2)  # (batch, head, time1, d_k)

            # (batch, head, time1, d_k)
            q_with_bias_u = (q + self.pos_bias_u).transpose(1, 2)
            # (batch, head, time1, d_k)
            q_with_bias_v = (q + self.pos_bias_v).transpose(1, 2)

            # compute attention score
            # first compute matrix a and matrix c
            # as described in https://arxiv.org/abs/1901.02860 Section 3.3
            # (batch, head, time1, time2)
            matrix_ac = torch.matmul(q_with_bias_u, k.transpose(-2, -1))

            # compute matrix b and matrix d
            # (batch, head, time1, time2)
            matrix_bd = torch.matmul(q_with_bias_v, p.transpose(-2, -1))
            matrix_bd = self.rel_shift(matrix_bd)
            # drops extra elements in the matrix_bd to match the matrix_ac's size
            matrix_bd = matrix_bd[:, :, :, : matrix_ac.size(-1)]

            scores = (matrix_ac + matrix_bd) / self.s_d_k  # (batch, head, time1, time2)

            out = self.forward_attention(v, scores, mask)
```

with sth like this:
```python
            q, k, v = self.forward_qkv(query, key, value)
            q = q.transpose(1, 2)  # (batch, time1, head, d_k)

            n_batch_pos = pos_emb.size(0)
            p = self.linear_pos(pos_emb).view(n_batch_pos, -1, self.h, self.d_k)
            p = p.transpose(1, 2)  # (batch, head, time1, d_k)

            # (batch, head, time1, d_k)
            q_with_bias_u = (q + self.pos_bias_u).transpose(1, 2)
            # (batch, head, time1, d_k)
            q_with_bias_v = (q + self.pos_bias_v).transpose(1, 2)

            # compute attention score
            # first compute matrix a and matrix c
            # as described in https://arxiv.org/abs/1901.02860 Section 3.3
            # (batch, head, time1, time2)
            #matrix_ac = torch.matmul(q_with_bias_u, k.transpose(-2, -1))

            # compute matrix b and matrix d
            # (batch, head, time1, time2)
            matrix_bd = torch.matmul(q_with_bias_v, p.transpose(-2, -1))
            matrix_bd = self.rel_shift(matrix_bd)
            # drops extra elements in the matrix_bd to match the matrix_ac's size
            matrix_bd = matrix_bd[:, :, :, : matrix_ac.size(-1)]

            #scores = (matrix_ac + matrix_bd) / self.s_d_k  # (batch, head, time1, time2)

            #out = self.forward_attention(v, scores, mask)

            out = torch.nn.functional.scaled_dot_product_attention(q_with_bias_u, k, v, attn_mask = matrix_bd)
```

?

Or is your argument that these computations of q_with_bias_u/q_with_bias_v/matrix_bd still slows down everything and does not bring any benefit from use of sdpa/mem_efficient_attn?

**Comment by vadimkantorov** [comment]
@desh2608 It seems that arbitrary attention mask is supported in mem_efficient_attention already, and will be better supported in FlexAttention backend in PyTorch soon. But for that, it would be nice to have the PR linked above reviewed by someone...
